### PR TITLE
Multi Visual Studio version support 

### PR DESCRIPTION
**Pull request by claraberendsen**
### Description
This PR restores the functionality of selecting which Visual Studio version you want the CI to run with.
This is a cherry-pick from the last approach of #784, for more debugging and lot's of test commits you can check the related PR.

This is paired to the changes to `ros2-cookbooks` in  https://github.com/ros-infrastructure/ros2-cookbooks/pull/74 

Changes

- Inserts a new key `@vs_version` in the json solo that specifies the VS release
- Modifies on job run the value according to CI_VS_VERSION
- Updates options to target 2019 and 2022 releases

### CODE REVIEW

**Comment by clalancette** [comment]
@claraberendsen @nuclearsandwich I think I finally, finally figured out all of the complicated issues going on with this PR.  Please see the long explanation in https://github.com/ros2/ci/commit/a31b09ac7677afa6aed290cb630287e5b297905d for why everything here is where it is.  I would very much appreciate a look from the both of you, since I'd like to get this in to unblock @ahcorde .

Since this PR only touches the Windows jobs, I'm only going to run CI for it.  But I'm going to run quite a lot of different CI here.  Note that all of these jobs will be run using the `test_ci_windows` job, since that is the one that has the changes for Windows in the job (once this is approved, I'll obviously do a full deploy).

* Rolling Windows Release 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/481/)
* Rolling Windows Release 2022 (expected to start, but fail compiling, since we need further changes) [Build Status](http://ci.ros2.org/job/test_ci_windows/479/)
* Rolling Windows Debug 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/482/)
* Rolling Windows Debug 2022 (expected to start, but fail compiling, since we need further changes) [Build Status](http://ci.ros2.org/job/test_ci_windows/480/)
* Jazzy Windows Release 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/484/)
* Jazzy Windows Debug 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/483/)
* Iron Windows Release 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/485/)
* Iron Windows Debug 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/486/)
* Humble Windows Release 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/487/)
* Humble Windows Debug 2019 [Build Status](http://ci.ros2.org/job/test_ci_windows/488/)

File: .gitmodules
Comment (claraberendsen) [inline .gitmodules:23]: Before merging this should be reverted to latest

File: ros2_batch_job/windows_batch/__init__.py
Comment (claraberendsen) [inline ros2_batch_job/windows_batch/__init__.py:49]: I think this is a fair compromise and removes the problem of docker CMD execution string expansion that has weird and unexpected behavior.

File: windows_docker_resources/Dockerfile
Comment (claraberendsen) [inline windows_docker_resources/Dockerfile:33]: Thanks for fixing my spelling error :heart:

**Review by claraberendsen — COMMENTED** [review]
I don't seem to have permission to officially approve this PR, but it looks good to me. 
Thanks @clalancette  for taking this up and pushing it to the finish line, I would like to get @nuclearsandwich review in before merging to have a fresh set of eyes that has not been hammering at this issue before to catch anything I might have missed (I might be too eager to not look at this for some time).

File: job_templates/ci_job.xml.em
Comment (nuclearsandwich) [inline job_templates/ci_job.xml.em:245]:
```suggestion
set SOLO_FILE=windows_docker_resources\install_ros2_%CI_ROS_DISTRO%.json
```

I understand the temptation since `DOCKERFILE` is right above, but I expect the word separator unless this was a `Solofile` or similar.

File: job_templates/ci_job.xml.em
Comment (nuclearsandwich) [inline job_templates/ci_job.xml.em:252]: Companion to an earlier suggestion.

```suggestion
powershell -noexit "(Get-Content ${Env:SOLO_FILE}).replace('@vs_version', ${Env:VISUAL_STUDIO_VERSION}) | Set-Content ${Env:SOLO_FILE}"
```

File: job_templates/packaging_job.xml.em
Comment (nuclearsandwich) [inline job_templates/packaging_job.xml.em:241]:
```suggestion
set SOLO_FILE=windows_docker_resources\install_ros2_%CI_ROS_DISTRO%.json
```

Companion to an earlier suggestion.

File: job_templates/packaging_job.xml.em
Comment (nuclearsandwich) [inline job_templates/packaging_job.xml.em:248]:
```suggestion
powershell -noexit "(Get-Content ${Env:SOLO_FILE}).replace('@vs_version', ${Env:VISUAL_STUDIO_VERSION}) | Set-Content ${Env:SOLO_FILE}"
```

Companion to an earlier suggestion.

File: job_templates/packaging_job.xml.em
Comment (nuclearsandwich) [inline job_templates/packaging_job.xml.em:247]:
```suggestion
rem "Change the chef-solo configuration file to set a specific the Visual Studio version.
```

I got hung up on 'solo install JSON file' so I attempted a re-word.

File: job_templates/ci_job.xml.em
Comment (nuclearsandwich) [inline job_templates/ci_job.xml.em:251]:
```suggestion
rem "Change the chef-solo configuration file to set a specific the Visual Studio version.
```

I got hung up on parsing 'solo install JSON file' so I attempted a re-word.

**Review by nuclearsandwich — APPROVED** [review]
A few minor suggestions but otherwise this is ready to go!

File: job_templates/ci_job.xml.em
Comment (clalancette) [reply job_templates/ci_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

File: job_templates/ci_job.xml.em
Comment (clalancette) [reply job_templates/ci_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

File: job_templates/packaging_job.xml.em
Comment (clalancette) [reply job_templates/packaging_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

File: job_templates/packaging_job.xml.em
Comment (clalancette) [reply job_templates/packaging_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

File: job_templates/packaging_job.xml.em
Comment (clalancette) [reply job_templates/packaging_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

File: job_templates/ci_job.xml.em
Comment (clalancette) [reply job_templates/ci_job.xml.em]: Fixed in [b7ae4a5](https://github.com/ros2/ci/pull/786/commits/b7ae4a59d079670a8f7a8772d5ba0b3102013732)

**Comment by clalancette** [comment]
I did a test deploy to https://ci.ros2.org/view/All/job/test_ci_windows/ , and then ran two jobs:

* [MSVC 2019](https://ci.ros2.org/view/All/job/test_ci_windows/501/)
* [MSVC 2022](https://ci.ros2.org/view/All/job/test_ci_windows/502)

As expected, the 2019 one completed, and the 2022 one failed to compile.  That's because we don't have our dependencies setup for 2022 yet.

With that, I'm going to go ahead and merge this in and deploy it for real.

**Comment by clalancette** [comment]
And this is now deployed.

**Pull request merged**
